src/impl_/pyclass_init.rs

```diff
@@ -17,69 +20,70 @@ pub trait PyObjectInit<T>: Sized + Sealed {
 
     #[doc(hidden)]
     fn can_be_subclassed(&self) -> bool;
 }
 
-/// Initializer for Python native types, like `PyDict`.
-pub struct PyNativeTypeInitializer<T: PyTypeInfo>(pub PhantomData<T>);
+/// Initializer for Python native types, like [PyDict].
+pub struct PyNativeTypeInitializer<T: PyTypeInfo + PyClassBaseType>(pub PhantomData<T>);
 
-impl<T: PyTypeInfo> PyObjectInit<T> for PyNativeTypeInitializer<T> {
+impl<T: PyTypeInfo + PyClassBaseType> PyObjectInit<T> for PyNativeTypeInitializer<T> {
+    /// call `__new__` ([ffi::PyTypeObject::tp_new]) for the native base type.
+    /// This will allocate a new python object and initialize the part relating to the native base type.
     unsafe fn into_new_object(
         self,
         py: Python<'_>,
         subtype: *mut PyTypeObject,
+        args: &Bound<'_, PyTuple>,
+        kwargs: Option<&Bound<'_, PyDict>>,
     ) -> PyResult<*mut ffi::PyObject> {
         unsafe fn inner(
             py: Python<'_>,
-            type_object: *mut PyTypeObject,
+            native_base_type: *mut PyTypeObject,
             subtype: *mut PyTypeObject,
+            args: &Bound<'_, PyTuple>,
+            kwargs: Option<&Bound<'_, PyDict>>,
+            override_tp_new: Option<ffi::newfunc>,
         ) -> PyResult<*mut ffi::PyObject> {
-            // HACK (due to FIXME below): PyBaseObject_Type's tp_new isn't happy with NULL arguments
-            let is_base_object = type_object == std::ptr::addr_of_mut!(ffi::PyBaseObject_Type);
-            let subtype_borrowed: Borrowed<'_, '_, PyType> = subtype
-                .cast::<ffi::PyObject>()
-                .assume_borrowed_unchecked(py)
-                .downcast_unchecked();
-
-            if is_base_object {
-                let alloc = subtype_borrowed
-                    .get_slot(TP_ALLOC)
-                    .unwrap_or(ffi::PyType_GenericAlloc);
-
-                let obj = alloc(subtype, 0);
-                return if obj.is_null() {
-                    Err(PyErr::fetch(py))
-                } else {
-                    Ok(obj)
-                };
-            }
+            let tp_new = if let Some(tp_new) = override_tp_new {
+                tp_new
+            } else {
+                native_base_type
+                    .cast::<ffi::PyObject>()
+                    .assume_borrowed_unchecked(py)
+                    .downcast_unchecked::<PyType>()
+                    .get_slot(TP_NEW)
+                    .ok_or_else(|| {
+                        PyTypeError::new_err("cannot construct type that does not define __new__")
+                    })?
+            };
 
-            #[cfg(Py_LIMITED_API)]
-            unreachable!("subclassing native types is not possible with the `abi3` feature");
+            let obj = tp_new(
+                subtype,
+                args.as_ptr(),
+                kwargs
+                    .map(|obj| obj.as_ptr())
+                    .unwrap_or(std::ptr::null_mut()),
+            );
```

Comment:
there is a limitation here, where in python a class has to call `super(MyClass, cls).__new__(cls, *args, **kwargs)` manually so it can choose which options to propagate to the super class. With the current approach the arguments propagate as-is